### PR TITLE
fix(langchain): tools in message history

### LINKED ISSUE #2938: 🐛 Bug Report: LangChain Tool handling. Tool content in message history and 'tool' message type
### Which component is this bug for?

Langchain Instrumentation

### 📜 Description

Larger bug: In message history in assistant messages (`gen_ai.prompt.N.content`) the content is empty if the message was a tool call.

Smaller bug: Message type `tool` automatically gets converted to `unknown`. I am not sure if any of the observability platforms (in particular, Traceloop) rely on the message type to be one of `user`, `assistant`, `unknown`, but I'd say conversions like this must happen in the backend.

### 👟 Reproduction steps

```
from langchain_core.messages import HumanMessage, AIMessage, ToolMessage, SystemMessage, BaseMessage
from langchain_openai import ChatOpenAI

def get_weather(location: str) -> str:
    return "sunny"

messages: list[BaseMessage] = [
    SystemMessage(content="Be crisp and friendly."),
    HumanMessage(content="Hey, what's the weather in San Francisco?"),
    AIMessage(
        content="",
        tool_calls=[
            {
                "name": "get_weather",
                "args": {"location": "San Francisco"},
                "id": "tool_123",
                "type": "tool_call",
            }
        ],
    ),
    ToolMessage(content="Sunny as always!", tool_call_id="tool_123"),
    HumanMessage(content="What's the weather in London?"),
]
model = ChatOpenAI(model="gpt-4.1-nano", temperature=0)
model_with_tools = model.bind_tools([get_weather])
result = model_with_tools.invoke(messages)
```

### 👍 Expected behavior

`gen_ai.prompts.2.content` has the tool_calls
`gen_ai.prompts.3.role == tool`

### 👎 Actual Behavior with Screenshots

`gen_ai.prompts.2.content` is empty
`gen_ai.prompts.3.role == unknown`

### 🤖 Python Version

3.13

### 📃 Provide any additional context for the Bug.

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [x] I checked and didn't find similar issue

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by dinmukhamedm**
Fixes #2938 

Screenshot 2025-05-12 at 23 28 29

- [X] I have added tests that cover my changes.
- [X] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [X] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

----

> [!IMPORTANT]
> Adds support for 'tool' message type and tool call handling in Langchain instrumentation with corresponding tests.
> 
>   - **Behavior**:
>     - Adds support for 'tool' message type in `_message_type_to_role()` in `callback_handler.py`.
>     - Updates `_set_chat_request()` in `callback_handler.py` to handle tool calls by checking `msg.content` and `tool_calls`.
>   - **Tests**:
>     - Adds `test_tool_calls_with_history.yaml` to test tool call interactions.
>     - Adds `test_tool_calls_with_history()` in `test_tool_calls.py` to verify tool call handling and span attributes.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=traceloop%2Fopenllmetry&utm_source=github&utm_medium=referral)<sup> for ec1ebdd98f0e252fb7b3eef359cd6c3bd6ff87b4. You can [customize](https://app.ellipsis.dev/traceloop/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py [hunk 1]
```diff
@@ -53,6 +53,8 @@ def _message_type_to_role(message_type: str) -> str:
         return "system"
     elif message_type == "ai":
         return "assistant"
+    elif message_type == "tool":
+        return "tool"
```
Comment (dinmukhamedm) [inline packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py:57]: @nirga I did discuss this in the issue. Hope this is fine with Traceloop's backend. Also, I think we already did something similar for Anthropic's `thinking` block in the output

**Comment by skull8888888** [comment]
Hey @nirga, can you please have a look at it, it's an extremely blocker for us. Thank you!

File: packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py [hunk 2]
```diff
@@ -148,8 +150,14 @@ def _set_chat_request(
                     f"{SpanAttributes.LLM_PROMPTS}.{i}.role",
                     _message_type_to_role(msg.type),
                 )
-                # if msg.content is string
-                if isinstance(msg.content, str):
+                tool_calls = dict(msg).get("tool_calls", msg.additional_kwargs.get("tool_calls"))
```
Comment (nirga) [inline packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py:153]: @dinmukhamedm shouldn't it be in `tool_calls` like here?
https://github.com/traceloop/openllmetry/blob/3a972381a06ace229e67985617d3d09983da6f25/packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py#L462

Comment (skull8888888) [reply packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py:153]: this PR is for the cases when input messages contain previous tool calls.

**Comment by dinmukhamedm** [comment]
@nirga also, while at it, I added handling a (not so) corner case where anthropic prepends a tool call with a text block in model output (see a couple new tests)

File: packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py (hunk 2 above)
Comment (nirga) [reply packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py]: Right, but should we send the tool calls in the same format as other instrumentations?

Comment (dinmukhamedm) [reply packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py]: Yeah, that makes sense. You mean like here? https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py#L393

That is `gen_ai.prompts.{i}.tool_calls.{j}.name`?

Comment (dinmukhamedm) [reply packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py]: done!

**Review by nirga — APPROVED** [review]
Thanks @dinmukhamedm 👑 
You should become a collaborator here :)
